cmd/runsim/main.go

```diff
@@ -195,119 +202,98 @@ wait:
 		}
 	}
 
-	if slackConfigSupplied() {
-		if checkIfLast() {
-			_, _, _ = pushLogs(logObjprefix, runsimLogfile)
-			slackMessage(slackToken, slackChannel, &slackThread, fmt.Sprintln("Simulation is finished!"))
-		}
-	}
-
-	// analyse results and exit with 1 on first error
+	// analyze results and collect the log file handles
 	close(results)
-	for rc := range results {
-		if !rc {
-			os.Exit(1)
+	var okSeeds, failedSeeds, exports []string
+	for seed := range results {
+		if seed.Failed {
+			failedSeeds = append(failedSeeds, seed.Stderr, seed.Stdout)
+		} else {
+			okSeeds = append(okSeeds, seed.Stderr, seed.Stdout)
 		}
+		exports = append(exports, seed.ExportParams, seed.ExportState)
 	}
-	os.Exit(0)
-}
 
-func buildCommand(testName, blocks, period, genesis string, seed int) string {
-	return fmt.Sprintf("go test %s -run %s -Enabled=true -NumBlocks=%s -Genesis=%s -Verbose=true " +
-		"-Commit=true -Seed=%d -Period=%s -ExportParamsPath %s -ExportStatePath %s -v -timeout 24h",
-		pkgName, testName, blocks, genesis, seed, period, ExportParamsPath, ExportStatePath)
-}
-
-func makeCmd(cmdStr string) *exec.Cmd {
-	cmdSlice := strings.Split(cmdStr, " ")
-	return exec.Command(cmdSlice[0], cmdSlice[1:]...)
-}
-
-func makeFilename(seed int) string {
-	return fmt.Sprintf("app-simulation-seed-%d-date-%s", seed, time.Now().Format("01-02-2006_150405"))
-}
+	if notifyGithub || notifySlack {
+		publishResults(okSeeds, failedSeeds, exports)
+	}
 
-func makeFailedSeedMsg(seed int, stdoutKey, stderrKey, bucket string, logsPushed bool) string {
-	if logsPushed {
-		return fmt.Sprintf("*Seed %s: FAILED*. *<https://%s.s3.amazonaws.com/%s|stdout>* *<https://%s.s3.amazonaws.com/%s|stderr>*\nTo reproduce run: ```\n%s\n```",
-			strconv.Itoa(seed), bucket, stdoutKey, bucket, stderrKey, buildCommand(testname, blocks, period, genesis, seed))
+	if len(failedSeeds) > 0 {
+		os.Exit(1)
 	}
-	return fmt.Sprintf("*Seed %s: FAILED*. \nTo reproduce run: ```\n%s\n```\n*Log upload failed:* ```\n%s\n```",
-		strconv.Itoa(seed), buildCommand(testname, blocks, period, genesis, seed), bucket)
+
+	os.Exit(0)
 }
 
-func worker(id int, seeds <-chan int) {
+func worker(id int, seeds <-chan *Seed, results chan *Seed) {
```

Comment:
As discussed in a slack chat, this is a cause of concern to me as this pointer is not goroutine safe. AFAIK @mircea-c is on it